Content.Server/Chunking/ChunkingSystem.cs

```diff
@@ -47,70 +50,59 @@ public override void Shutdown()
 
+        var enlargement = viewEnlargement ?? chunkSize;
+        AddViewerChunks(player, chunks, indexPool, chunkSize, enlargement);
         foreach (var uid in session.ViewSubscriptions)
         {
-            viewers.Add(uid);
+            AddViewerChunks(uid, chunks, indexPool, chunkSize, enlargement);
         }
 
-        return viewers;
+        return chunks;
     }
 
-    private Dictionary<NetEntity, HashSet<Vector2i>> GetChunksForViewers(
-        HashSet<EntityUid> viewers,
-        int chunkSize,
+    private void AddViewerChunks(EntityUid viewer,
+        Dictionary<NetEntity, HashSet<Vector2i>> chunks,
         ObjectPool<HashSet<Vector2i>> indexPool,
-        ObjectPool<Dictionary<NetEntity, HashSet<Vector2i>>> viewerPool,
+        int chunkSize,
         float viewEnlargement)
     {
-        var chunks = viewerPool.Get();
-        DebugTools.Assert(chunks.Count == 0);
+        if (!_xformQuery.TryGetComponent(viewer, out var xform))
+            return;
+
+        var pos = _transform.GetWorldPosition(xform);
+        var bounds = _baseViewBounds.Translated(pos).Enlarged(viewEnlargement);
 
-        foreach (var viewerUid in viewers)
+        var state = (chunks, indexPool, chunkSize, bounds);
+        _mapManager.FindGridsIntersecting(xform.MapID, bounds, ref state, AddGridChunks, true);
+    }
+
+    private bool AddGridChunks(
+        EntityUid uid,
+        MapGridComponent grid,
+        ref (Dictionary<NetEntity, HashSet<Vector2i>>, ObjectPool<HashSet<Vector2i>>, int, Box2) state)
+    {
+        var netGrid = GetNetEntity(uid);
+
+        var (chunks, pool, size, bounds) = state;
+        if (!chunks.TryGetValue(netGrid, out var set))
         {
-            if (!_xformQuery.TryGetComponent(viewerUid, out var xform))
-            {
-                Log.Error($"Player has deleted viewer entities? Viewers: {string.Join(", ", viewers.Select(ToPrettyString))}");
-                continue;
-            }
-
-            var pos = _transform.GetWorldPosition(xform);
-            var bounds = _baseViewBounds.Translated(pos).Enlarged(viewEnlargement);
-            var grids = new List<Entity<MapGridComponent>>();
-            _mapManager.FindGridsIntersecting(xform.MapID, bounds, ref grids, true);
-
-            foreach (var grid in grids)
-            {
-                var netGrid = GetNetEntity(grid);
-
-                if (!chunks.TryGetValue(netGrid, out var set))
-                {
-                    chunks[netGrid] = set = indexPool.Get();
-                    DebugTools.Assert(set.Count == 0);
-                }
-
-                var enumerator = new ChunkIndicesEnumerator(_transform.GetInvWorldMatrix(grid).TransformBox(bounds), chunkSize);
-
-                while (enumerator.MoveNext(out var indices))
-                {
-                    set.Add(indices.Value);
-                }
-            }
+            chunks[netGrid] = set = pool.Get();
+            DebugTools.Assert(set.Count == 0);
```

Comment:
Won't this allocate as you're getting a reference to EntityManager that isn't passed into the method?